File: 222. Count Complete Tree Nodes/Rust/Count Complete Tree Nodes.rs
use std::rc::Rc;
use std::cell::RefCell;

type Node = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn count_nodes(root: Node) -> i32 
    {
        fn dfs(node: &Node) -> i32
        {
            if let Some(node) = node
            {
                return 1 + dfs(&node.borrow().left) + dfs(&node.borrow().right);
            } else { return 0; }
        }

        dfs(&root)
    }
}

File: Problems & Solutions/222. Count Complete Tree Nodes/15 November 2022/Java/Count Complete Tree Nodes.java
class Solution {
    public int countNodes(TreeNode root) 
    {
        if (root == null) return 0;

        TreeNode left = root, right = root;
        int l = 1, r = 1;
        
        while ((left  = left.left)   != null) ++l;
        while ((right = right.right) != null) ++r;
        
        if (l == r) return (1 << l) - 1;
        
        return 1 + countNodes(root.left) + countNodes(root.right);
    }
}